File: functions/fitregmodel.m
% 
% FITREGMODEL Fits a parameter-free distance distribution to one (or several)
%             time-domain signals, using regularization
%
%   P = FITREGMODEL(V,K,r,regtype,alpha)
%   Regularization of the N-point signal (V) to a M-point distance
%   distribution (P) given a M-point distance axis (r) and NxM point kernel
%   (K). The regularization parameter (alpha) controls the regularization 
%   properties.
%
%   [P,Pci] = FITREGMODEL(V,K,r,regtype,method)
%   An estimation of the 95%-confidence intervals is returned as a second 
%   output (Pci). If more than one confidence level is requested, (Pci)
%   is returned as a cell array containing the confidence intervals at the 
%   different confidence levels.
%
%   [P,Pci] = FITREGMODEL(V,K,r,regtype,method)
%   Instead of passing a numerial value for the regularization parameter
%   (alpha), the name of a selection method (method) can be passed and 
%   the regularization parameter will be automatically selected by means
%   of the selregparam function.
%
%   [P,Pci,alpha,stats] = FITREGMODEL(V,K,r,regtype,method)
%   The value of the regularization parameter found via selection
%   methods and used for fitting (P) is returned in (alpha).  A structure 
%   containing different statistical estimators of goodness of fit is returned as (stats). 
%
%   The type of regularization employed in FITREGMODEL is set by the regtype
%   input argument. The regularization models implemented in FITREGMODEL are:
%          'tikhonov' -   Tikhonov regularization
%          'tv'       -   Total variation regularization
%          'huber'    -   pseudo-Huber regularization
%
%   P = FITREGMODEL({V1,V2,...},{K1,K2,...},r,regtype,alpha)
%   Passing multiple signals/kernels enables global fitting of the
%   a single-distribution model to all data. The global fit weights
%   are automatically computed according to their contribution to ill-posedness.
%
%   P = FITREGMODEL(...,'Property',Values)
%   Additional (optional) arguments can be passed as property-value pairs.
%
% The properties to be passed as options can be set in any order. 
%
%   'Solver' - Solver to be used to solve the minimization problems
%                      'fnnls' - Fast non-negative least-squares
%                      'lsqnonneg' - Non-negative least-squares 
%                      'fmincon' - Non-linear constrained minimization
%                      'bppnnls' -  Block principal pivoting non-negative least-squares solver
%   'NonNegConstrained' - Enable/disable non-negativity constraint (true/false)
%   'HuberParam' - Huber parameter used in the 'huber' model (default = 1.35).
%   'GlobalWeights' - Array of weighting coefficients for the individual signals in
%                     global fitting.
%   'RegOrder' - Order of the regularization operator L (default = 2)
%   'ConfidenceLevel' - Confidence lvel(s) for the confidence intervals (0-1, default=0.95)
%   'TolFun' - Optimizer function tolerance
%   'MaxIter' - Maximum number of optimizer iterations
%   'MaxFunEvals' - Maximum number of optimizer function evaluations
%   'Verbose' - Display options for the solvers:
%                    'off' - no information displayed  
%                    'final' - display solver exit message
%                    'iter-detailed' - display state of solver at each iteration                   
%                     See MATLAB doc optimoptions for detailed explanation

% This file is a part of DeerLab. License is MIT (see LICENSE.md). 
% Copyright(c) 2019-2020: Luis Fabregas, Stefan Stoll and other contributors.


function [P,Pci,alpha,stats] = fitregmodel(V,K,r,RegType,alpha,varargin)

% Turn off warnings to avoid ill-conditioned warnings 
warning('off','MATLAB:nearlySingularMatrix')

%--------------------------------------------------------------------------
% Parse & Validate Required Input
%--------------------------------------------------------------------------
if nargin<3
    error('Not enough input arguments.')
end
if nargin<4 || isempty(RegType)
    RegType = 'tikhonov';
elseif isa(RegType,'function_handle')
    RegFunctional = RegType;
    RegType = 'custom';
else
    validateattributes(RegType,{'char'},{'nonempty'})
    allowedInput = {'tikhonov','tv','huber','custom'};
    RegType = validatestring(RegType,allowedInput);
end
if  nargin<5 || isempty(alpha)
   alpha = 'aic'; 
end

% Check if user requested some options via name-value input
optionalProperties = {'TolFun','Solver','NonNegConstrained','Verbose','MaxFunEvals','MaxIter','HuberParam','GlobalWeights','RegOrder','ConfidenceLevel','internal::parseLater'};
[TolFun,Solver,NonNegConstrained,Verbose,MaxFunEvals,MaxIter,HuberParam,GlobalWeights,RegOrder,ConfidenceLevel] ...
    = parseoptional(optionalProperties,varargin);


% Remove used options from varargin so they are not passed to selregparam
for i=1:numel(optionalProperties)
    Idx = find(cellfun(@(x)(ischar(x) && strcmpi(x,optionalProperties{i})),varargin));
    varargin(Idx:Idx+1) = [];
end

if strcmp(RegType,'custom')
    GradObj = false;
else
    GradObj = true;
end
if isa(alpha,'char')
    alpha = selregparam(V,K,r,RegType,alpha,[{'GlobalWeights'},{GlobalWeights},varargin]);
else
    validateattributes(alpha,{'numeric'},{'scalar','nonempty','nonnegative'},mfilename,'RegParam')
end
% validateattributes(r,{'numeric'},{'nonempty','increasing','nonnegative'},mfilename,'r')
validateattributes(r,{'numeric'},{'nonempty','nonnegative'},mfilename,'r')

%--------------------------------------------------------------------------
% Parse & Validate Optional Input
%--------------------------------------------------------------------------

if isempty(Verbose)
    Verbose = 'off';
else
    validateattributes(Verbose,{'char'},{'nonempty'},mfilename,'Verbose')
end
if isempty(ConfidenceLevel)
   ConfidenceLevel = 0.95; 
else
    validateattributes(ConfidenceLevel,{'numeric'},{'nonnegative','nonempty'},mfilename,'ConfidenceLevel')
    if any(ConfidenceLevel<=0 | ConfidenceLevel>1)
       error('The ''ConfidenceLevel'' option must contain values in the range [0 1].') 
    end
end
if isempty(RegOrder)
    RegOrder = 2;
else
    validateattributes(RegOrder,{'numeric'},{'scalar','nonnegative'})
end
if isempty(TolFun)
    TolFun = 1e-9;
else
    validateattributes(TolFun,{'numeric'},{'scalar','nonempty','nonnegative'},'regularize','nonNegLSQsolTol')
end
if isempty(Solver) && ~strcmp(RegType,'custom')
    Solver = 'fnnls';
elseif isempty(Solver) && strcmp(RegType,'custom')
        Solver = 'fmincon';
else
    validateattributes(Solver,{'char'},{'nonempty'})
    allowedInput = {'analytical','fnnls','lsqnonneg','bppnnls','fmincon'};
    Solver = validatestring(Solver,allowedInput);
end

if isempty(MaxIter)
    MaxIter = 2e7;
else
    validateattributes(MaxIter,{'numeric'},{'scalar','nonempty'},mfilename,'MaxIter')
end

if isempty(HuberParam)
    HuberParam = 1.35;
else
    validateattributes(HuberParam,{'numeric'},{'scalar','nonempty','nonnegative'},mfilename,'MaxFunEvals')
end

if isempty(MaxFunEvals)
    MaxFunEvals = 2e7;
else
    validateattributes(MaxFunEvals,{'numeric'},{'scalar','nonempty'},mfilename,'MaxFunEvals')
end

if isempty(NonNegConstrained)
    NonNegConstrained = true;
else
    validateattributes(NonNegConstrained,{'logical'},{'nonempty'},'regularize','NonNegConstrained')
end
if ~iscell(V)
    V = {V};
end
if ~iscell(K)
    K = {K};
end
if ~isempty(GlobalWeights)
    validateattributes(GlobalWeights,{'numeric'},{'nonnegative'})
    if length(GlobalWeights) ~= length(V)
        error('The same number of global fit weights as signals must be passed.')
    end
    % Normalize weights
    GlobalWeights = GlobalWeights/sum(GlobalWeights);
end
if numel(K)~=numel(V)
    error('The number of kernels must be equal to the number of kernels.')
end
for i = 1:numel(V)
    if ~iscolumn(V{i})
        V{i} = V{i}.';
    end
    if ~isreal(V{i})
        V{i} = real(V{i});
    end
    if length(V{i})~=size(K{i},1)
        error('The number of rows in K must match the number of elements in V.')
    end
    validateattributes(V{i},{'numeric'},{'nonempty'},mfilename,'S')
end
if nargout>1
    getConfidenceIntervals = true;
else
    getConfidenceIntervals = false;
end
if strcmp(RegType,'custom')
    getConfidenceIntervals = false;
end
%--------------------------------------------------------------------------
% Regularization processing
%--------------------------------------------------------------------------

nr = size(K{1},2);
L = regoperator(nr,RegOrder);
InitialGuess = zeros(nr,1);

dr = mean(diff(r));

% If unconstrained regularization is requested then solve analytically
if ~NonNegConstrained && ~strcmp(Solver,'fmincon')
    Solver = 'analytical';
end

% If using LSQ-based solvers then precompute the KtK and KtS input arguments
if ~strcmp(Solver,'fmincon') || getConfidenceIntervals
    [Q,KtS,weights] =  lsqcomponents(V,r,K,L,alpha,RegType,HuberParam,GlobalWeights);
end

% Solve the regularization functional minimization problem
switch lower(Solver)
    
    case 'analytical'
        P = zeros(nr,1);
        for i = 1:length(V)
            PseudoInverse = Q\K{i}.';
            P = P + weights(i)*PseudoInverse*V{i};
        end
        
    case 'lsqnonneg'
        solverOpts = optimset('Display','off','TolX',TolFun);
        P = lsqnonneg(Q,KtS,solverOpts);

    case 'fnnls'
        [P,~,~,flag] = fnnls(Q,KtS,InitialGuess,TolFun,Verbose);
        %In some cases, fnnls may return negatives if tolerance is to high
        if flag==-1
            %... in those cases continue from current solution
            [P,~,~,flag] = fnnls(Q,KtS,P,1e-20);
        end
        if flag==-2
            warning('FNNLS cannot solve the problem. Regularization parameter may be too large.')
        end
        
    case 'bppnnls'
        P = nnls_bpp(Q,KtS,Q\KtS);
        
    case 'fmincon'
        % Constrained Tikhonov/Total variation/Huber regularization
        if NonNegConstrained
            NonNegConst = zeros(nr,1);
        else
            NonNegConst = [];
        end
        if ~strcmp(RegType,'custom')
            RegFunctional = regfunctional(RegType,V,L,K,alpha,HuberParam);
        else
            % Parse errors in the analyzed function, and reformat them
            RegFunctional = @(P)errorhandler(RegFunctional,'regfcn',P);
        end
        
        fminconOptions = optimoptions(@fmincon,'SpecifyObjectiveGradient',GradObj,'MaxFunEvals',MaxFunEvals,'Display',Verbose,'MaxIter',MaxIter);
        [P,~,exitflag] =  fmincon(RegFunctional,InitialGuess,[],[],[],[],NonNegConst,[],[],fminconOptions);
        % Check how optimization exited...
        if exitflag == 0
            %... if maxIter exceeded (flag=0) then double iterations and continue from where it stopped
            fminconOptions = optimoptions(fminconOptions,'MaxIter',2*MaxIter,'MaxFunEvals',2*MaxFunEvals);
            P  = fmincon(RegFunctional,P,[],[],[],[],NonNegConst,[],[],fminconOptions);
        end
end

% Normalize distribution
P = P/sum(P)/dr;

% Calculate confidence intervals
%-------------------------------------------------------------------------------
if getConfidenceIntervals
    
    % Estimate the contribution to P variance from the different signals
    sigP = 0;
    for i = 1:numel(V)
        % Estimate the residual standard deviation
        sig = std(V{i} - K{i}*P);
        
        % Get the regularized pseudoinverse
        Q = lsqcomponents(V{i},r,K{i},L,alpha,RegType,HuberParam);
        pKinv = Q\K{i}.';
        
        % Get standard error from covariance matrix
        covP = pKinv*pKinv.';
        sigP_ = sig*sqrt(diag(covP));
        
        sigP = sigP + weights(i)*sigP_;
    end
    
    % Get the Gaussian quantile according to requested coverage
    p = 1 - (1 - ConfidenceLevel)/2;
    
    Pci = cell(numel(p),1);
    %Get the confidence intervals at the requested confidence levels
    for j=1:numel(p)
        % Get the critical value of corresponding normal distribution
        z = norm_inv(p(j));
        % Compute the standard confidence intervals, constrained to be nonnegative
        Pci{j}(:,1) = max(P - z*sigP,0);
        Pci{j}(:,2) = P + z*sigP;
    end
    
    %Do not return a cell if only one confidence level is requested
    if numel(p)==1
        Pci = Pci{1};
    end
end

% If requested compute Goodness of Fit for all signals
if nargout>3
    stats = cell(numel(V),1);
    for i=1:numel(V)
        Vfit = K{i}*P;
        Ndof = numel(V{i});
        stats{i} = gof(V{i},Vfit,Ndof);
    end
    
    if numel(V)==1
        stats = stats{1};
    end
end

% Turn warnings back on
warning('on','MATLAB:nearlySingularMatrix')

end

File: deerlab/fitregmodel.py
# fitregmodel.py - Fiting of regularization problems
# --------------------------------------------------
# This file is a part of DeerLab. License is MIT (see LICENSE.md).
# Copyright(c) 2019-2021: Luis Fabregas, Stefan Stoll and other contributors.

import numpy as np
import matplotlib.pyplot as plt
from deerlab.nnls import fnnls, cvxnnls, nnlsbpp
import deerlab as dl
import copy
from deerlab.utils import hccm, goodness_of_fit
from deerlab.classes import UQResult, FitResult

def fitregmodel(V, K, r, regtype='tikhonov', regparam='aic', regorder=2, solver='cvx', 
                weights=None, huberparam=1.35, nonnegativity=True, obir=False,
                uq=True, renormalize=True, noiselevelaim=None, tol=None, maxiter=None):
    r"""
    Fits a non-parametric distance distribution to one (or several) signals using regularization aproaches 
    via non-negative least-squares (NNLS).

    Parameters 
    ----------
    V : array_like or list of array_like  
        Dipolar signal, multiple datasets can be globally evaluated by passing a list of signals.
    
    K : 2D-array_like array or list of 2D-array_like
        Dipolar kernel, if a list of signals is specified, a corresponding list of kernels must be passed as well.
    
    r : array_like
        Distance axis, in nanometers.
    
    regtype : string, optional
        Regularization functional type: 
    
        * ``'tikhonov'`` - Tikhonov regularizaton
        * ``'tv'``  - Total variation regularization
        * ``'huber'`` - Huber regularization
        
        The default is ``'tikhonov'``.   
    
    regparam : string or scalar, optional
        Method for the automatic selection of the optimal regularization parameter:
    
        * ``'lr'`` - L-curve minimum-radius method (LR)
        * ``'lc'`` - L-curve maximum-curvature method (LC)
        * ``'cv'`` - Cross validation (CV)
        * ``'gcv'`` - Generalized Cross Validation (GCV)
        * ``'rgcv'`` - Robust Generalized Cross Validation (rGCV)
        * ``'srgcv'`` - Strong Robust Generalized Cross Validation (srGCV)
        * ``'aic'`` - Akaike information criterion (AIC)
        * ``'bic'`` - Bayesian information criterion (BIC)
        * ``'aicc'`` - Corrected Akaike information criterion (AICC)
        * ``'rm'`` - Residual method (RM)
        * ``'ee'`` - Extrapolated Error (EE)          
        * ``'ncp'`` - Normalized Cumulative Periodogram (NCP)
        * ``'gml'`` - Generalized Maximum Likelihood (GML)
        * ``'mcl'`` - Mallows' C_L (MCL)
        
        The regularization parameter can be manually specified by passing a scalar value instead of a string.
        The default ``'aic'``.
    
    weights : array_like, optional
        Array of weighting coefficients for the individual signals in global fitting. 
        If not specified all datasets are weighted inversely proportional to their noise levels.
    
    regorder : int scalar, optional
        Order of the regularization operator, the default is 2.
    
    solver : string, optional
        Optimizer used to solve the non-negative least-squares problem: 

        * ``'cvx'`` - Optimization of the NNLS problem using cvxopt
        * ``'fnnls'`` - Optimization using the fast NNLS algorithm.
        * ``'nnlsbpp'`` - Optimization using the block principal pivoting NNLS algorithm.
        
        The default is ``'cvx'``.

    uq : boolean, optional
        Enable/disable the uncertainty quantification analysis, by default it is enabled. 
    
    nonnegativity : boolean, optional
        Enforces the non-negativity constraint on computed distance distributions, by default it is enabled.
    
    huberparam : scalar, optional
        Value of the Huber parameter used in Huber regularization, the default is 1.35.
    
    renormalize : boolean, optional
        Enable/disable renormalization of the fitted distribution, by default it is enabled.
    
    obir : boolean, optional
        Enable/disable the use of the Osher-Bregman iterated regularization algorithm, by default it is disabled.
    
    noiselevelaim : scalar, optional
        Noise level at which to stop the OBIR algorithm. If not specified it is automatically estimated from the fit residuals.

    tol : scalar, optional 
        Tolerance value for convergence of the NNLS algorithm. If not specified (``None``), the value is set automatically to ``tol = max(K.T@K.shape)*norm(K.T@K,1)*eps``.
        It is only valid for the ``'cvx'`` and ``'fnnls'`` solvers, it has no effect on the ``'nnlsbpp'`` solver.
        
    maxiter: scalar, optional  
        Maximum number of iterations before termination. If not specified (``None``), the value is set automatically to ``maxiter = 5*shape(K.T@K)[1]``.
        It is only valid for the ``'cvx'`` and ``'fnnls'`` solvers, it has no effect on the ``'nnlsbpp'`` solver.

    Returns
    -------
    :ref:`FitResult` with the following fields defined:
    P : ndarray
        Fitted distance distribution.
    
    V : ndarray 
        Fitted dipolar signal.

    Puncert : :ref:`UQResult`
        Covariance-based uncertainty quantification of the fitted distance distribution.
   
    VUncert : :ref:`UQResult`
        Covariance-based uncertainty quantification of the fitted dipolar signal.
     
    regparam : float int
        Regularization parameter used in the optimization
    
    scale : float int or list of float int
        Amplitude scale(s) of the dipolar signal(s).
    
    plot : callable
        Function to display the results. It will 
        display the fitted signals and distance distributions with
        confidence intervals. The function returns the figure object 
        (``matplotlib.figure.Figure``) object as output, which can be 
        modified. Using ``fig = plot(show=False)`` will not render
        the figure unless ``display(fig)`` is called. 
    
    stats : dict
        Goodness of fit statistical estimators (if ``full_output=True``):

        * ``stats['chi2red']`` - Reduced \chi^2 test
        * ``stats['r2']`` - R^2 test
        * ``stats['rmsd']`` - Root-mean squared deviation (RMSD)
        * ``stats['aic']`` - Akaike information criterion
        * ``stats['aicc']`` - Corrected Akaike information criterion
        * ``stats['bic']`` - Bayesian information criterion
    
    success : bool
        Whether or not the optimizer exited successfully.
    
    cost : float
        Value of the cost function at the solution.
    
    residuals : ndarray
        Vector of residuals at the solution.

    """
    # Prepare signals, kernels and weights if multiple are passed
    V, K, weights, subsets = dl.utils.parse_multidatasets(V, K, weights, precondition=False)

    if len(subsets)>1:
        prescales = [max(V[subset]) for subset in subsets]
        for scale,subset in zip(prescales,subsets):
            V[subset] = V[subset]/scale
    else:
        prescales = [1]



    # Determine an optimal value of the regularization parameter if requested
    if type(regparam) is str:
        alpha = dl.selregparam(V,K,r,regtype,regparam,regorder=regorder,weights=weights, nonnegativity=nonnegativity,huberparam=huberparam)
    else:
        alpha = regparam

    # Prepare components of the LSQ problem
    L = dl.regoperator(r,regorder)
    KtKreg, KtV = dl.lsqcomponents(V,K,L,alpha,weights, regtype=regtype, huberparam=huberparam)

    # Determine the type of problem to solve
    if obir:
        problem = 'obir'
    elif nonnegativity:
        problem = 'nnls'
    else:
        problem = 'unconstrained'

    # Solve unconstrained LSQ problem
    if problem == 'unconstrained':
        Pfit = np.linalg.solve(KtKreg,KtV)

    # Solve using Osher-Bregman iterated regularization
    elif problem == 'obir':
        Pfit = _obir(V,K,L,regtype,alpha,weights,noiselevelaim=noiselevelaim,huberparam=huberparam, solver=solver)
    
    # Solve non-negative LSQ problem
    elif problem == 'nnls':

        if solver == 'fnnls':
            Pfit = fnnls(KtKreg,KtV,tol=tol,maxiter=maxiter)
        elif solver == 'nnlsbpp':
            Pfit = nnlsbpp(KtKreg,KtV,np.linalg.solve(KtKreg,KtV))
        elif solver == 'cvx':
            Pfit = cvxnnls(KtKreg,KtV,tol=tol,maxiter=maxiter)
        else:
            raise KeyError(f'{solver} is not a known non-negative least squares solver.')

    # Get fit final status
    Vfit = K@Pfit
    success = ~np.all(Pfit==0)

    # Construct residual parts for for the residual and regularization terms
    res = weights*(V - K@Pfit)

    # Construct Jacobians for the residual and penalty terms
    Jres = K*weights[:,np.newaxis]
    res,J = _augment(res,Jres,regtype,alpha,L,Pfit,huberparam)

    # Get objective function value
    fval = np.linalg.norm(res)**2

    # Uncertainty quantification
    # ----------------------------------------------------------------
    if uq:
        # Calculate the heteroscedasticity consistent covariance matrix 
        covmat = hccm(J,res,'HC1')
        
        # Construct confidence interval structure for P
        if nonnegativity:
            NonNegConst = np.zeros(len(r))
        else: 
            NonNegConst = None
        Puq = UQResult('covariance',Pfit,covmat,NonNegConst)
    else:
        Puq = UQResult('void')

    # Get fitted signals and their uncertainty
    # --------------------------------------
    modelfit, modelfituq = [],[]
    for subset in subsets: 
        subset_model = lambda P: (K@P)[subset]
        modelfit.append(subset_model(Pfit))
        if uq: 
            modelfituq.append(Puq.propagate(subset_model))
        else:
            modelfituq.append(UQResult('void'))
    
    # Re-normalization of the distributions
    # --------------------------------------
    postscale = np.trapz(Pfit,r)
    if renormalize:
        Pfit = Pfit/postscale
        if uq:
            Puq_ = copy.deepcopy(Puq) # need a copy to avoid infite recursion on next step
            Puq = Puq_.propagate(lambda P: P/postscale, lbm=np.zeros_like(Pfit))

    # Goodness-of-fit
    # --------------------------------------
    H = K@(np.linalg.pinv(KtKreg)@K.T)
    stats = []
    for subset in subsets: 
        Ndof = len(V[subset]) - np.trace(H)
        stats.append(goodness_of_fit(V[subset],Vfit[subset],Ndof))


    # Signal amplitude scales
    # ---------------------------------------
    scales = []
    for i in range(len(subsets)): 
        scales.append(prescales[i]*postscale)

    if len(scales)==1:
        scales = scales[0]
        stats = stats[0]
        modelfit = modelfit[0]
        modelfituq = modelfituq[0]

    # Results display
    # ---------------------------------------
    def plotfcn(show=False):
        fig = _plot(subsets,V,Vfit,r,Pfit,Puq,show)
        return fig 

    return FitResult(P=Pfit, V=modelfit, Puncert=Puq, Vuncert=modelfituq, regparam=alpha, scale=scales, stats=stats, 
                     plot=plotfcn, cost=fval, residuals=res, success=success)
# ===========================================================================================


def _augment(res,J,regtype,alpha,L,P,eta):
# ===========================================================================================
    """ 
    LSQ residual and Jacobian augmentation
    =======================================

    Augments the residual and the Jacobian of a LSQ problem to include the
    regularization penalty. The residual and Jacobian contributions of the 
    specific regularization methods are analytically introduced. 
    """
    eps = np.finfo(float).eps
    # Compute the regularization penalty augmentation for the residual and the Jacobian
    if regtype == 'tikhonov':
        resreg = L@P
        Jreg = L
    elif regtype == 'tv':
        resreg =((L@P)**2 + eps)**(1/4)
        Jreg = 2/4*((( ( (L@P)**2 + eps)**(-3/4) )*(L@P))[:, np.newaxis]*L)
    elif regtype == 'huber':
        resreg = np.sqrt(np.sqrt((L@P/eta)**2 + 1) - 1)
        Jreg = 0.5/(eta**2)*( (((np.sqrt((L@P/eta)**2 + 1) - 1 + eps)**(-1/2)*(((L@P/eta)**2 + 1+ eps)**(-1/2)))*(L@P))[:, np.newaxis]*L )

    # Include regularization parameter
    resreg = alpha*resreg
    Jreg = alpha*Jreg

    # Augment jacobian and residual
    res = np.concatenate((res,resreg))
    J = np.concatenate((J,Jreg))

    return res,J
# ===========================================================================================


def _obir(V,K,L, regtype, alpha, weights, noiselevelaim=None, huberparam=1.35 , solver = 'cvx'):
# ===========================================================================================
    """
    Osher's Bregman-iterated regularization method
    ==============================================

    P = OBIR(V,K,r,regtype,alpha)

    OBIR of the N-point signal (V) to a M-point distance
    distribution (P) given a M-point distance axis (r) and NxM point kernel
    (K). The regularization parameter (alpha) controls the regularization
    properties.

    The type of regularization employed in OBIR is set by the regtype
    input argument. The regularization models implemented in OBIR are:
        'tikhonov' -   Tikhonov regularization
        'tv'       -   Total variation regularization
        'huber'    -   pseudo-Huber regularization

    Additional (optional) keyword arguments:
      weights          weights  
      noiselevelaim    the noise level to aim for, automatic if set to None
      huberparam       parameter for Huber reguarization
      solver           solver to use: 'cvx', 'fnnls', or 'nnlsbpp'
    """

    if noiselevelaim is None:
        noiselevelaim = dl.noiselevel(V,'der')

    MaxOuterIter = 5000
    stopDivergent = False

    # Preparation
    #-------------------------------------------------------------------------------

    # Initialize
    nr = np.shape(L)[1]
    subGrad = np.zeros(nr)
    Counter = 1
    Iteration = 1
    Pfit = np.zeros(nr)

    # Osher's Bregman Iterated Algorithm
    #-------------------------------------------------------------------------------
    diverged = np.zeros(len(V))
    semiconverged = np.zeros(len(V))

    # Precompute the KtK and KtS input arguments
    [KtKreg,KtV] = dl.lsqcomponents(V,K,L,alpha, weights, regtype=regtype, huberparam=huberparam)

    while Iteration <= MaxOuterIter:

        # Store previous iteration distribution
        Pprev = Pfit
        
        # Update
        KtVsg = KtV - subGrad
        
        # Get solution of current Bregman iteration
        if solver == 'fnnls':
            Pfit = fnnls(KtKreg,KtVsg)
        elif solver == 'nnlsbpp':
            Pfit = nnlsbpp(KtKreg,KtVsg,np.linalg.solve(KtKreg,KtVsg))
        elif solver == 'cvx':
            Pfit = cvxnnls(KtKreg, KtVsg)
        else:
            raise KeyError(f'{solver} is not a known non-negative least squares solver.')
        
        # Update subgradient at current solution
        subGrad = subGrad + weights*K.T@(K@Pfit - V)
        
        # Iteration control
        #--------------------------------------------------------------------------
        if Iteration == 1:
            # If at first iteration, the residual deviation is already below the
            # noise deviation then impose oversmoothing and remain at first iteration
            if noiselevelaim  > np.std(K@Pfit - V):
                alpha = alpha*2**Counter
                Counter = Counter + 1
                
                # Recompute the KtK and KtS input arguments with new alpha
                KtKreg,KtV = dl.lsqcomponents(V,K,L,alpha, weights, regtype=regtype, huberparam=huberparam)
            else:
                # Once the residual deviation is above the threshold, then proceed
                # further with the Bregman iterations
                Iteration  = Iteration + 1
        else:
            # For the rest of the Bregman iterations control the condition and stop
            # when fulfilled
            diverged = np.std(K@Pprev - V) < np.std(K@Pfit - V)
            semiconverged = noiselevelaim > np.std(K@Pfit - V)

            if semiconverged:
                break
            else:
                Iteration  = Iteration + 1

            # If residual deviation starts to diverge, stop
            if stopDivergent and diverged:
                Pfit = Pprev
                break

    return Pfit
# ===========================================================================================

def _plot(subsets,Vexp,Vfit,r,Pfit,Puq,show=False):
# ===========================================================================================
    nSignals = len(subsets)
    fig,axs = plt.subplots(nSignals+1,figsize=[7,3+3*nSignals])
    for i in range(nSignals): 
        subset = subsets[i]
        # Plot the experimental signal and fit
        axs[i].plot(Vexp[subset],'.',color='grey',alpha=0.5)
        axs[i].plot(Vfit[subset],'tab:blue')
        axs[i].grid(alpha=0.3)
        axs[i].set_xlabel('Array Elements')
        axs[i].set_ylabel(f'V[{i}]')
        axs[i].legend(('Data','Fit'))

    # Get confidence intervals for the distance distribution
    Pci95 = Puq.ci(95)
    Pci50 = Puq.ci(50)
    # Plot the distribution
    axs[nSignals].plot(r,Pfit,'tab:blue')
    axs[nSignals].fill_between(r,Pci95[:,0], Pci95[:,1],facecolor='tab:blue',linestyle='None',alpha=0.2)
    axs[nSignals].fill_between(r,Pci50[:,0], Pci50[:,1],facecolor='tab:blue',linestyle='None',alpha=0.4)
    axs[nSignals].set_xlabel('Distance (nm)')
    axs[nSignals].set_ylabel('P (nm⁻¹)')
    axs[nSignals].legend(('Fit','95%-CI','50%-CI'))
    axs[nSignals].grid(alpha=0.3)
    plt.tight_layout()
    if show:
        plt.show()
    else:
        plt.close()
    return fig
# ===========================================================================================
